Use more POSIX thread-safe functions in libunix

MisterDA opened the pull request:
- Use thread-safe functions [`getgrnam_r`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/getgrnam_r.html), [`getgrgid_r`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/getgrgid_r.html), [`getpwnam_r`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/getpwnam_r.html), [`getpwuid_r`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/getpwuid_r.html), introduced in POSIX-2008, if available.

  This sample program prints 4096 on macOS and 1024 on Arch Linux in my tests.

  ```c
  #include <unistd.h>
  #include <stdio.h>

  int main(void) {
      long int initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
      printf("%ld\n", initlen);
      return 0;
  }
  ```

- Use thread-safe functions [`gmtime_r`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/gmtime_r.html), [`localtime_r`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/localtime_r.html), [`getlogin_r`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/getlogin_r.html) and if available.

- Fix `mktime` error-checking. `(time_t) -1` is a valid value. Check the error as suggested by the C23 draft and POSIX.

MisterDA committed:
Use POSIX thread-safe getgrnam_r, getgrgid_r, getpwnam_r, getpwuid_r

- https://pubs.opengroup.org/onlinepubs/9799919799/functions/getgrnam_r.html
- https://pubs.opengroup.org/onlinepubs/9799919799/functions/getgrgid_r.html
- https://pubs.opengroup.org/onlinepubs/9799919799/functions/getpwnam_r.html
- https://pubs.opengroup.org/onlinepubs/9799919799/functions/getpwuid_r.html

MisterDA committed:
Use POSIX thread-safe gmtime_r and localtime_r

- https://pubs.opengroup.org/onlinepubs/9799919799/functions/gmtime_r.html
- https://pubs.opengroup.org/onlinepubs/9799919799/functions/localtime_r.html

MisterDA committed:
Fix mktime error condition

- https://pubs.opengroup.org/onlinepubs/9799919799/functions/mktime.html
- https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/mktime-mktime32-mktime64?view=msvc-170
- C11 standard § 7.26.2.3 The mktime function

MisterDA committed:
Update changes

MisterDA committed:
Use POSIX thread-safe getlogin_r

dustanddreams commented on otherlibs/unix/getgr.c:
```diff
@@ -43,31 +46,77 @@ static value alloc_group_entry(struct group *entry)
   if (! caml_string_is_c_safe(name)) caml_raise_not_found();
+
+#ifdef HAVE_GETGRNAM_R
+  long int initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
```
```suggestion
  long initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
```
Did you paste this from an example in the spec?

MisterDA replied:
I did take some inspiration it, yes.

dustanddreams commented on otherlibs/unix/getgr.c:
```diff
@@ -43,31 +46,77 @@ static value alloc_group_entry(struct group *entry)
+
+#ifdef HAVE_GETGRNAM_R
+  long int initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
+  size_t len = initlen == -1 ? /* default */ 1024 : (size_t) initlen;
```
It doesn't hurt to write
```suggestion
  size_t len = initlen <= 0 ? /* default */ 1024 : (size_t) initlen;
```
which will appease static analyzers complaining about the risk of casting a negative `long` value into an unsigned `size_t` value.

MisterDA replied:
Thanks!

dustanddreams commented on otherlibs/unix/getgr.c:
```diff
@@ -18,10 +18,13 @@
+#include <unistd.h>
 #include <grp.h>
 
+#if !defined(HAVE_GETGRNAM_R) || !defined(HAVE_GETGRID_R)
+#include <errno.h>
+#endif
```
That's wrong. The `_r` codepath does not refer to `errno` but nevertheless uses `ERANGE` which comes from `errno.h`, so you should always include that file instead of relying upon it being included by another header.

```suggestion
#include <errno.h>
```

MisterDA replied:
Right, fixed.

dustanddreams commented on otherlibs/unix/getgr.c:
```diff
@@ -43,31 +46,77 @@ static value alloc_group_entry(struct group *entry)
+  struct group *resultp;
+
+#ifdef HAVE_GETGRGID_R
+  long int initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
```
Same as above.
```suggestion
  long initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
```

dustanddreams commented on otherlibs/unix/getgr.c:
```diff
@@ -43,31 +46,77 @@ static value alloc_group_entry(struct group *entry)
+
+#ifdef HAVE_GETGRGID_R
+  long int initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
+  size_t len = initlen == -1 ? /* default */ 1024 : (size_t) initlen;
```
Same suggestion as above.
```suggestion
  size_t len = initlen <= 0 ? /* default */ 1024 : (size_t) initlen;
```

MisterDA commented:
Thanks for the review. I've applied your suggestions, and also added a change to use `getlogin_r`.

dra27 commented:
Cross-referencing #10505

smuenzel commented on otherlibs/unix/getlogin.c:
```diff
@@ -13,17 +13,36 @@
+  long initlen = sysconf(_SC_LOGIN_NAME_MAX);
+  bufsize = initlen <= 0 ? _POSIX_LOGIN_NAME_MAX : (size_t) initlen;
+#endif
+  name = caml_stat_alloc(initlen);
```
Are all combinations of `#ifdef` exercised in tests? It seems to me that if `HAVE_GETLOGIN_R` and `LOGIN_NAME_MAX` is defined, then `initlen` will not be defined on this line

MisterDA replied:
Thanks for the review! The problem is that it should be `bufsize` here instead of `initlen`.

dra27 commented on otherlibs/unix/gmtime.c:
```diff
@@ -81,11 +81,12 @@ CAMLprim value caml_unix_mktime(value t)
-  tm.tm_wday = Int_val(Field(t, 6));
-  tm.tm_yday = Int_val(Field(t, 7));
+  tm.tm_wday = -1; /* tm.tm_wday = Int_val(Field(t, 6)); */
+  tm.tm_yday = -1; /* tm.tm_yday = Int_val(Field(t, 7)); */
```
Why are you setting both?

MisterDA replied:
Setting `tm_wday = -1` to catch errors and `tm_isdst = -1` because its undefined should be enough, thanks.

dra27 commented on configure.ac:
```diff
@@ -2349,10 +2349,7 @@ AC_CHECK_FUNC([gettimeofday],
-
-AC_CHECK_FUNC([mktime], [AC_DEFINE([HAS_MKTIME], [1])])
-AC_CHECK_FUNCS_ONCE([gmtime_r localtime_r])
+AC_CHECK_FUNCS_ONCE([mktime gmtime_r localtime_r])
```
I realise we end up with a slightly weird `HAS_MKTIME` and `HAVE_GMTIME`, but it's not like they're used right next to each other!

MisterDA replied:
Reverted to keeping `HAS_MKTIME`, and not using `HAVE_MKTIME`.

dra27 commented on otherlibs/unix/getgr.c:
```diff
@@ -43,31 +43,71 @@ static value alloc_group_entry(struct group *entry)
+  int e;
+  while ((e = getgrnam_r(String_val(name), &result, buffer, len, &resultp))
+         == ERANGE) {
+    len *= 2;
```
I think we usually put a cap on this, just to stop anything ridiculous from possibly happening?

dra27 replied:
Obviously these 3 comments apply equally to all four functions!

MisterDA replied:
I've added an arbitrary cap...

dra27 commented on otherlibs/unix/getgr.c:
```diff
@@ -43,31 +43,71 @@ static value alloc_group_entry(struct group *entry)
+  long initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
+  size_t len = initlen <= 0 ? /* default */ 1024 : (size_t) initlen;
+  struct group result;
+  char *buffer = caml_stat_alloc(len);
```
This may be pushing this too far - and I haven't checked what we do with the other `_r` functions, but this path with `getgrnam_r` can in theory raise `Out_of_memory` where the other can't. Would it be better to be using the non-raising versions and instead raising `Unix_error` with `ENOMEM` if a malloc/realloc fails?

MisterDA replied:
Moved to `_noexc` variants and raising unix errors.

dra27 commented on otherlibs/unix/getgr.c:
```diff
@@ -43,31 +43,71 @@ static value alloc_group_entry(struct group *entry)
 {
-  struct group * entry;
+  value res;
+  struct group *resultp;
```
Mainly out of curiosity, why change the name from `entry` to `resultp`?

MisterDA replied:
To match posix examples. Reverted.

dra27 commented on otherlibs/unix/gmtime.c:
```diff
@@ -39,22 +39,30 @@ static value alloc_tm(struct tm *tm)
-  tm = gmtime(&clock);
-  if (tm == NULL) caml_unix_error(EINVAL, "gmtime", Nothing);
-  return alloc_tm(tm);
+  time_t timer = (time_t) Double_val(t);
```
Again, I'm not totally convinced the variable renaming was necessary?

MisterDA replied:
Reverted.

dra27 commented on otherlibs/unix/getlogin.c:
```diff
@@ -13,17 +13,36 @@
+  bufsize = LOGIN_NAME_MAX;
+#else
+  long initlen = sysconf(_SC_LOGIN_NAME_MAX);
+  bufsize = initlen <= 0 ? _POSIX_LOGIN_NAME_MAX : (size_t) initlen;
```
Similar memory comment as for `getgrnam_r`, etc.

MisterDA replied:
Done, thanks.

dra27 reviewed (commented):
Slightly repetitively, because I've just (finally!) reviewed a series of them, but I'd prefer not to have these changes that necessitate `compatibility.h` getting bigger.

I'm surprised, but happy to be corrected at my surprise at `resultp` vs `presult` (or just leaving the names as they were!!).

The question of `Out_of_memory` vs `ENOMEM` will be highly academic until it suddenly happens (like discovering that you do actually get users who are members of more than 32 groups.......), so that's probably worth thinking about now while there's a bunch of these functions being added.

I will confess that I did wonder while reviewing the fourth or fifth similar looking pattern between `foo` and `foo_r` whether it could be affectively written using macros, but I shall now go and atone for that in a confessional...

MisterDA commented:
Thanks for the review! I've applied your suggestions and fixed a few issues on the way.

> Slightly repetitively, because I've just (finally!) reviewed a series of them, but I'd prefer not to have these changes that necessitate `compatibility.h` getting bigger.

I've reverted this change.

> I'm surprised, but happy to be corrected at my surprise at `resultp` vs `presult` (or just leaving the names as they were!!).

POSIX uses `resultp` or `xxxptr` in their examples or declarations. I've reverted the renaming.

> The question of `Out_of_memory` vs `ENOMEM` will be highly academic until it suddenly happens (like discovering that you do actually get users who are members of more than 32 groups.......), so that's probably worth thinking about now while there's a bunch of these functions being added.

I've switched to `_noexc` functions and raising unix errors with `ENOMEM`, but I've added somewhat arbitrary limits on the length. Shout now if something isn't right!

> I will confess that I did wonder while reviewing the fourth or fifth similar looking pattern between `foo` and `foo_r` whether it could be affectively written using macros, but I shall now go and atone for that in a confessional...

Yeah... or an OCaml typed GADT-based DSL that generates C code!

dra27 reviewed (approved):
LGTM, thanks - hopefully at some point soon we'll go with Issue 7, which would mean the non-`_r` versions can then be retired!

dra27 commented:
Going through [precheck#1097](https://ci.inria.fr/ocaml/job/precheck/1097/) to check the BSDs, etc.
